src/backend/joanie/badges/providers/obf.py

```diff
@@ -0,0 +1,468 @@
+"""OpenBadgeFactory provider."""
+
+import json
+import logging
+import re
+from collections.abc import Iterable
+from functools import cached_property
+from typing import Literal, Optional
+from urllib.parse import urljoin
+
+# pylint: disable=no-name-in-module
+import requests
+from pydantic import (
+    BaseModel,
+    EmailStr,
+    HttpUrl,
+    NoneStr,
+    ValidationError,
+    root_validator,
+)
+
+from ..exceptions import AuthenticationError, BadgeProviderError
+from .base import BaseProvider
+
+logger = logging.getLogger(__name__)
+
+
+class OAuth2AccessToken(requests.auth.AuthBase):
+    """Add OAuth2 access token to HTTP API requests header."""
+
+    def __init__(self, access_token):
+        """Instantiate requests Auth object with generated access_token."""
+
+        self.access_token = access_token
+
+    def __call__(self, request):
+        """Modify and return the request."""
+
+        request.headers.update(
+            {
+                "Authorization": f"Bearer {self.access_token}",
+            }
+        )
+
+        return request
+
+
+class OBFAPIClient(requests.Session):
+    """Open Badge Factory API Client."""
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        *args,
+        raise_for_status: bool = False,
+        **kwargs,
+    ):
+        """Override default requests.Session instantiation to handle authentication."""
+
+        super().__init__(*args, **kwargs)
+
+        self.api_root_url = "https://openbadgefactory.com"
+        self.api_version_prefix = "v1"
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.raise_for_status = raise_for_status
+        self.auth = self._get_auth
+        self.headers.update(
+            {
+                "Content-Type": "application/json",
+            }
+        )
+
+    @cached_property
```

Comment:
Are you sure `@cached_property` serves your purpose here as the cached value relies on the class instance, so a request to OBF will be send for each new `OBFAPIClient` instance.

Reply:
We can bet that a script or management command execution will rely on a single API client instance for a limited amount of time (the script execution time), hence, caching this token avoids renewing it at each API request. What alternative approach would you suggest?

Reply:
After chatting with @sampaccoud, I now understand that you expect a more reliable cache to rely on in HTTP requests processing context. Hence, I refactored this method as a LRU cached static method. Hope this was what you expected.

Reply:
This is exactly what I expected 👌